stl/inc/algorithm

```diff
@@ -5118,6 +5118,155 @@ _FwdIt shift_right(_ExPo&&, _FwdIt _First, _FwdIt _Last, _Iter_diff_t<_FwdIt> _P
+                    --_UEnd_at;
+                }
+            }
+
+            _Seek_wrapped(_First, _RANGES _Move_backward_common(_UFirst, _UEnd_at, _ULast));
+            return {_First, _RANGES next(_First, _Last)};
+        } else if constexpr (sized_sentinel_for<_Se, _It> && bidirectional_iterator<_Se>) {
+            auto _Size = _ULast - _UFirst;
+            if (_Pos_to_shift >= _Size) {
+                _RANGES advance(_First, _Size);
+                return {_First, _First};
+            }
+
+            auto _UEnd_at = _RANGES next(_UFirst, _Size - _Pos_to_shift);
+            _Seek_wrapped(_First, _RANGES _Move_backward_common(_UFirst, _UEnd_at, _ULast));
+            return {_First, _RANGES next(_First, _Size)};
+        } else {
+            auto _UResult = _UFirst;
+
+            for (; 0 < _Pos_to_shift; --_Pos_to_shift) {
+                if (_UResult == _ULast) {
+                    _Seek_wrapped(_First, _UResult);
+                    return {_First, _First};
+                }
+                ++_UResult;
+            }
+            _Seek_wrapped(_First, _UResult);
+
+            auto _Trail = _UFirst;
+            auto _Lead  = _UResult;
+
+            for (; _Trail != _UResult; ++_Trail, (void) ++_Lead) {
+                if (_Lead == _ULast) {
+                    _RANGES _Move_unchecked(_UFirst, _Trail, _UResult);
+
+                    auto _WLead = _First;
+                    _Seek_wrapped(_WLead, _Lead);
+
+                    return {_First, _WLead};
+                }
+            }
+
+            // Here, _Trail = _UFirst + original _Pos_to_shift
+            // Here, _Lead = _UFirst + 2 * original _Pos_to_shift
+
+            for (;;) {
+                // This loop swaps the range [_UFirst, _UResult) with [_Trail, _Lead),
+                // advancing _Trail and _Lead by _Pos_to_shift
+                for (auto _Mid = _UFirst; _Mid != _UResult; ++_Mid, (void) ++_Trail, ++_Lead) {
+                    if (_Lead == _ULast) {
+                        _Trail = _RANGES _Move_unchecked(_Mid, _UResult, _Trail).out;
+                        _RANGES _Move_unchecked(_UFirst, _Mid, _Trail);
+
+                        auto _WLead = _First;
+                        _Seek_wrapped(_WLead, _Lead);
+
+                        return {_First, _WLead};
+                    }
+                    _RANGES iter_swap(_Mid, _Trail);
+                }
+            }
+        }
+    }
+
+    template <permutable _It, sentinel_for<_It> _Se>
+    constexpr subrange<_It> shift_right(_It _First, _Se _Last, _Iter_diff_t<_It> _Pos_to_shift) {
+        _Adl_verify_range(_First, _Last);
+        return _Shift_right(_First, _Last, _Pos_to_shift);
+    }
+
+    template <forward_range _Rng>
+        requires permutable<iterator_t<_Rng>>
+    constexpr borrowed_subrange_t<_Rng> shift_right(_Rng&& _Range, range_difference_t<_Rng> _Pos_to_shift) {
+        return _Shift_right(_RANGES begin(_Range), _RANGES end(_Range), _Pos_to_shift);

```

Comment:
This could use `_Ubegin` and `_Uend` instead of `_RANGES begin` and `_RANGES end`.